File: resource/temp.py
import math
import numpy as np
from math import cos, sin, sqrt, atan2, acos, pi, log10


def sph2cart1(r, th, phi):
    x = r * cos(phi) * sin(th)
    y = r * sin(phi) * sin(th)
    z = r * cos(th)

    return x, y, z


def cart2sph1(x, y, z):
    r = sqrt(x ** 2 + y ** 2 + z ** 2) + 1e-15
    th = acos(z / r)
    phi = atan2(y, x)

    return r, th, phi

def PatchFunction(thetaInDeg, phiInDeg, Freq, W, L, h, Er):
    """
    Taken from Design_patchr
    Calculates total E-field pattern for patch as a function of theta and phi
    Patch is assumed to be resonating in the (TMx 010) mode.
    E-field is parallel to x-axis
    W......Width of patch (m)
    L......Length of patch (m)
    h......Substrate thickness (m)
    Er.....Dielectric constant of substrate
    Refrence C.A. Balanis 2nd Edition Page 745
    """
    lamba = 3e8 / Freq

    theta_in = math.radians(thetaInDeg)
    phi_in = math.radians(phiInDeg)

    ko = 2 * math.pi / lamba

    xff, yff, zff = sph2cart1(999, theta_in, phi_in)                            # Rotate coords 90 deg about x-axis to match array_utils coord system with coord system used in the model.
    xffd = zff
    yffd = xff
    zffd = yff
    r, thp, php = cart2sph1(xffd, yffd, zffd)
    phi = php
    theta = thp

    if theta == 0:
        theta = 1e-9                                                              # Trap potential division by zero warning

    if phi == 0:
        phi = 1e-9

    Ereff = ((Er + 1) / 2) + ((Er - 1) / 2) * (1 + 12 * (h / W)) ** -0.5        # Calculate effictive dielectric constant for microstrip line of width W on dielectric material of constant Er

    F1 = (Ereff + 0.3) * (W / h + 0.264)                                        # Calculate increase length dL of patch length L due to fringing fields at each end, giving total effective length Leff = L + 2*dL
    F2 = (Ereff - 0.258) * (W / h + 0.8)
    dL = h * 0.412 * (F1 / F2)

    Leff = L + 2 * dL

    Weff = W                                                                    # Calculate effective width Weff for patch, uses standard Er value.
    heff = h * sqrt(Er)

    # Patch pattern function of theta and phi, note the theta and phi for the function are defined differently to theta_in and phi_in
    Numtr2 = sin(ko * heff * cos(phi) / 2)
    Demtr2 = (ko * heff * cos(phi)) / 2
    Fphi = (Numtr2 / Demtr2) * cos((ko * Leff / 2) * sin(phi))

    Numtr1 = sin((ko * heff / 2) * sin(theta))
    Demtr1 = ((ko * heff / 2) * sin(theta))
    Numtr1a = sin((ko * Weff / 2) * cos(theta))
    Demtr1a = ((ko * Weff / 2) * cos(theta))
    Ftheta = ((Numtr1 * Numtr1a) / (Demtr1 * Demtr1a)) * sin(theta)

    # Due to groundplane, function is only valid for theta values :   0 < theta < 90   for all phi
    # Modify pattern for theta values close to 90 to give smooth roll-off, standard model truncates H-plane at theta=90.
    # PatEdgeSF has value=1 except at theta close to 90 where it drops (proportional to 1/x^2) to 0

    rolloff_factor = 0.5                                                       # 1=sharp, 0=softer
    theta_in_deg = theta_in * 180 / math.pi                                          # theta_in in Deg
    F1 = 1 / (((rolloff_factor * (abs(theta_in_deg) - 90)) ** 2) + 0.001)       # intermediate calc
    PatEdgeSF = 1 / (F1 + 1)                                                    # Pattern scaling factor

    UNF = 1.0006                                                                # Unity normalisation factor for element pattern

    if theta_in <= math.pi / 2:
        Etot = Ftheta * Fphi * PatEdgeSF * UNF                                   # Total pattern by pattern multiplication
    else:
        Etot = 0

    return Etot

def GetPatchFields(PhiStart, PhiStop, ThetaStart, ThetaStop, Freq, W, L, h, Er):
    """"
    Calculates the E-field for range of thetaStart-thetaStop and phiStart-phiStop
    Returning a numpy array of form - fields[phiDeg][thetaDeg] = eField
    W......Width of patch (m)
    L......Length of patch (m)
    h......Substrate thickness (m)
    Er.....Dielectric constant of substrate
    """
    fields = np.ones((PhiStop, ThetaStop))                                      # Create initial array to hold e-fields for each position

    for phiDeg in range(PhiStart, PhiStop):
            for thetaDeg in range(ThetaStart, ThetaStop):                       # Iterate over all Phi/Theta combinations
                eField = PatchFunction(thetaDeg, phiDeg, Freq, W, L, h, Er)     # Calculate the field for current Phi, Theta
                fields[phiDeg][thetaDeg] = eField                               # Update array with e-field

    return fields

def PatchEHPlanePlot(Freq, W, L, h, Er, isLog=True):
    """
    Plot 2D plots showing E-field for E-plane (phi = 0) and the H-plane (phi = 90).
    """

    fields = GetPatchFields(0, 360, 0, 90, Freq, W, L, h, Er)


PatchEHPlanePlot((24 * 10 ^ 8), 0.03, 0.02, 0.0016, 4.4)

File: resource/temp.js
var f = 2.4 * Math.pow(10, 9)

var toMM = Math.pow(10, -3)

var pw = 23.700674815509178 * toMM
var pl = 21.587764623402762 * toMM
var fw = 9.480269926203672 * toMM
var fl = 12.760735864844596 * toMM
var effl = 0.02552147172968919 * toMM
var ele_l = 61.29393570460074
var ed = 2.156006183008272 * toMM
var gl = 58.34850048824735 * toMM
var gw = 57.18094474171285 * toMM
var h = 1.6 * toMM
var er = 4.4

var Fields = GetPatchFields(0, 360, 0, 90)

//console.log('fields', Fields)

phiSize = 360
thetaSize = 90

X = Array(phiSize).fill(Array(thetaSize).fill(1));
Y = Array(phiSize).fill(Array(thetaSize).fill(1));
Z = Array(phiSize).fill(Array(thetaSize).fill(1));

for (var phi_temp =0; phi_temp < phiSize; phi_temp++)
{
    for (var theta_temp =0; theta_temp < thetaSize; theta_temp++)
    {
        var e_temp = Fields[phi_temp][theta_temp]
        var t_temp = toRadians(theta_temp)
        var p_temp = toRadians(phi_temp)
        var ax = sph2cart1 (e_temp, t_temp, p_temp)
        X[phi_temp][theta_temp] = ax[0]
        Y[phi_temp][theta_temp] = ax[1]
        Z[phi_temp][theta_temp] = ax[2]
    }
}

var layout = {
    title: 'Surface Plot of EH Plane',
    scene: {
        xaxis: {
            gridcolor: 'rgb(255, 255, 255)',
            zerolinecolor: 'rgb(255, 255, 255)',
            showbackground: true,
            backgroundcolor: 'rgb(230, 230,230)'
        },
        yaxis: {
            gridcolor: 'rgb(255, 255, 255)',
            zerolinecolor: 'rgb(255, 255, 255)',
            showbackground: true,
            backgroundcolor: 'rgb(230, 230,230)'
        },
        zaxis: {
            gridcolor: 'rgb(255, 255, 255)',
            zerolinecolor: 'rgb(255, 255, 255)',
            showbackground: true,
            backgroundcolor: 'rgb(230, 230,230)'
        }
    }
}

var data = [{
        x: X,
        y: Y,
        z: Z,
        type: 'surface'
}]

function showPlot(){
    document.getElementById('plotd').display='block';
    Plotly.newPlot('plotd', data=data, {}, {});
}

function GetPatchFields(PhiStart, PhiStop, ThetaStart, ThetaStop)

{
    fields = Array(PhiStop).fill(Array(ThetaStop).fill(1));

    for (var phiDeg = PhiStart; phiDeg < PhiStop; phiDeg++) {
        for (var thetaDeg = ThetaStart; thetaDeg < ThetaStop; thetaDeg ++){
            var eField = PatchFunction (thetaDeg, phiDeg)
            fields[phiDeg][thetaDeg] = eField
        }
    }

    return fields
}

function toRadians (angle) {
  return angle * (Math.PI / 180);
}

function sph2cart1(r, th, phi){
  x = r * Math.cos(phi) * Math.sin(th)
  y = r * Math.sin(phi) * Math.sin(th)
  z = r * Math.cos(th)
  return [x, y, z]
}

function cart2sph1(x, y, z){
  r = Math.sqrt( Math.pow(x,2) + Math.pow(y,2) + Math.pow(z,2)) + Math.pow(10, -15)
  th = Math.acos(z / r)
  phi = Math.atan2(y, x)
  return [r, th, phi]
}

function PatchFunction (thetaInDeg, phiInDeg){

    var lambda = 3 * Math.pow(10, 8) / f;

    var theta_in = toRadians(thetaInDeg)
    var phi_in = toRadians(phiInDeg)

    var ko = 2 * Math.PI / lambda

    ff = sph2cart1(999, theta_in, phi_in)
    out = cart2sph1(ff[2], ff[0], ff[1])
    r = out[0]; theta = out[1]; phi = out[2];

    if (theta == 0)
       theta = Math.pow(10, -9)

    if (phi == 0)
        phi = Math.pow(10, -9)

    Ereff = ((er + 1) / 2) + ((er - 1) / 2) * (1 + 12 * (h / pw))
    Ereff = Math.pow(Ereff, -0.5)

    F1 = (Ereff + 0.3) * (pw / h + 0.264)
    F2 = (Ereff - 0.258) * (pw / h + 0.8)
    dL = h * 0.412 * (F1 / F2)

    Leff = pl + 2 * dL

    Weff = pw
    heff = h * Math.sqrt(er)

    Numtr2 = Math.sin(ko * heff * Math.cos(phi) / 2)
    Demtr2 = (ko * heff * Math.cos(phi)) / 2
    Fphi = (Numtr2 / Demtr2) * Math.cos((ko * Leff / 2) * Math.sin(phi))

    Numtr1 = Math.sin((ko * heff / 2) * Math.sin(theta))
    Demtr1 = ((ko * heff / 2) * Math.sin(theta))
    Numtr1a = Math.sin((ko * Weff / 2) * Math.cos(theta))
    Demtr1a = ((ko * Weff / 2) * Math.cos(theta))

    Ftheta = ((Numtr1 * Numtr1a) / (Demtr1 * Demtr1a)) * Math.sin(theta)

    rolloff_factor = 0.5
    theta_in_deg = theta_in * 180 / Math.PI
    F1 = 1 / (((rolloff_factor * (Math.abs(theta_in_deg) - 90)) ** 2) + 0.001)
    PatEdgeSF = 1 / (F1 + 1)

    UNF = 1.0006

    if (theta_in <= Math.PI / 2)
        Etot = Ftheta * Fphi * PatEdgeSF * UNF
    else
        Etot = 0

    return Etot
}